test/hotspot/jtreg/runtime/os/TestTrimNative.java

```diff
@@ -0,0 +1,279 @@
+        long trimInterval = 500; // twice per second
+        long ms1 = System.currentTimeMillis();
+        OutputAnalyzer output = runTestWithOptions (
+                new String[] { "-XX:+UnlockExperimentalVMOptions",
+                               "-XX:+TrimNativeHeap",
+                               "-XX:TrimNativeHeapInterval=" + trimInterval },
+                new String[] { "RUN", "5000" }
+        );
+        long ms2 = System.currentTimeMillis();
+        long runtime_ms = ms2 - ms1;
+
+        checkExpectedLogMessages(output, true, 500);
+
+        long maxTrimsExpected = runtime_ms / trimInterval;
+        long minTrimsExpected = maxTrimsExpected / 2;
+        parseOutputAndLookForNegativeTrim(output, (int)minTrimsExpected, (int)maxTrimsExpected);
+    }
+
+    // Test that a high trim interval effectively disables trimming
+    static private final void testHighTrimInterval() throws IOException {
+        OutputAnalyzer output = runTestWithOptions (
+                new String[] { "-XX:+UnlockExperimentalVMOptions",
+                               "-XX:+TrimNativeHeap",
+                               "-XX:TrimNativeHeapInterval=" + Integer.MAX_VALUE },
+                new String[] { "RUN", "5000" }
+        );
+        checkExpectedLogMessages(output, true, Integer.MAX_VALUE);
+        parseOutputAndLookForNegativeTrim(output,0, /*  minTrimsExpected */ 0  /*  maxTrimsExpected */);
+    }
+
+    // Test that trim-native gets disabled on platforms that don't support it.
+    static private final void testOffOnNonCompliantPlatforms() throws IOException {
+        OutputAnalyzer output = runTestWithOptions (
+                new String[] { "-XX:+UnlockExperimentalVMOptions",
+                               "-XX:+TrimNativeHeap" },
+                new String[] { "RUN", "0" }
+        );
+        checkExpectedLogMessages(output, false, 0);
+        output.shouldContain("Native trim not supported on this platform");
+    }
+
+    // Test trim native is disabled if explicitly switched off
+    static private final void testOffExplicit() throws IOException {
+        OutputAnalyzer output = runTestWithOptions (
+                new String[] { "-XX:+UnlockExperimentalVMOptions",
+                               "-XX:-TrimNativeHeap"
+                },
+                new String[] { "RUN", "0" }
+                );
+        checkExpectedLogMessages(output, false, 0);
+    }
+
+    // Test trim native is disabled if explicitly switched off
+    static private final void testOffByDefault() throws IOException {
+        OutputAnalyzer output = runTestWithOptions (null, new String[] { "RUN", "0" } );
+        checkExpectedLogMessages(output, false, 0);
+    }
+
+    public static void main(String[] args) throws Exception {
+
+        if (args.length == 0) {
+            throw new RuntimeException("Argument error");
+        }
+
+        if (args[0].equals("RUN")) {
```

Comment:
The usual trick is to pull this "internal" main into a class, and then reference it. Like:

```
public Test {
 
  public void test() {
    runWith("...", "Test$Workload");
  }  

  public class Workload {
    public static void main(String... args) {} 
  }
}
```